---
layout: default
title: Home
nav_order: 1
description: Identification, validation, and quantification of amino acid substitutions that arise from alternate RNA decoding using LC-MS proteomics data | Slavov Laboratory and single-cell proteomics center
permalink: /
---
{% include social-media-links.html %}

# Alternate RNA decoding

{: .fs-7 .fw-300}
**Beyond the genetic code:** Alternate translation diversify mammalian proteomes

## Quantifying amino acid substitutions from alternate RNA decoding

&nbsp;


[Decode article][Decode_article]{: .btn .fs-5 .mb-4 .mb-md-0 }
[Decode code @ GitHub][decode_Code]{: .btn .fs-5 .mb-4 .mb-md-0 }

------------

Amino acid substitutions may substantially alter protein stability and function, but the contribution of substitutions arising from alternate translation (deviations from the genetic code) is unknown. To explore it, we analyzed deep proteomic and transcriptomic data from over 1,000 human samples, including 6 cancer types and 26 healthy human tissues. This global analysis identified 60,024 high confidence substitutions corresponding to 8,801 unique sites in 1,990 proteins. Some substitutions are shared across samples, while others exhibit strong tissue-type and cancer specificity. Surprisingly, products of alternate translation are more abundant than their canonical counterparts for hundreds of proteins, suggesting sense codon recoding. Recoded proteins include transcription factors, proteases, signaling proteins, and proteins associated with neurodegeneration. Mechanisms contributing to substitution abundance include protein stability, codon frequency, codon-anticodon mismatches, and RNA modifications. We characterize sequence motifs around alternatively translated amino acids and how substitution ratios vary across protein domains, tissue types and cancers. Both the sequence and the tissue-specificity of alternatively translated proteins are conserved between human and mouse. These results demonstrate the contribution of alternate translation to diversifying mammalian proteomes, and its association with protein stability, tissue-specific proteomes, and diseases.  

* Tsour, S., Machne, R., *et al.* Alternate RNA decoding results in stable and abundant proteins in mammals. <!---*Nat Biotechnol* (2022). [10.1038/s41587-022-01389-w][plexDIA_Nature],  [Preprint][plexDIA_Article]--->


------------


<!---[![plexDIA: Multiplexed data-independent acquisition for increasing proteomics throughput](https://scp.slavovlab.net/Figs/plexDIA_4.png){: width="100%" .center-image}](https://scp.slavovlab.net/plexDIA) --->


------------



<!--- ## Perspectives on high-throughput multiplexed proteomics
* [Strategies for increasing the depth and throughput of protein analysis by plexDIA](https://pubs.acs.org/doi/10.1021/acs.jproteome.2c00721)  *Journal of Proteome Research*
* [Framework for multiplicative scaling of single-cell proteomics](https://www.nature.com/articles/s41587-022-01411-1), *Nature Biotechnology*
* [Increasing proteomics throughput](https://www.nature.com/articles/s41587-021-00881-z), *Nature Biotechnology*
* [Driving Single Cell Proteomics Forward with Innovation](https://pubmed.ncbi.nlm.nih.gov/34597050/), *Journal of Proteome Research*
* [Scaling up single-cell proteomics](https://doi.org/10.1016/j.mcpro.2021.100179), *Molecular and Cellular Proteomics* --->



## About the project

decode is a project developed in the [Slavov Laboratory](https://slavovlab.net) at [Northeastern University](https://www.northeastern.edu/). It was authored by [Shira Tsour](https://slavovlab.net/people.htm), [Rainer Machne](https://slavovlab.net/people.htm) [Andrew Leduc](http://andrewdleduc.com/), [Simon Widmer] and [Nikolai Slavov](https://coe.northeastern.edu/people/slavov-nikolai/).   


Contact the authors by email: [nslavov\{at\}northeastern.edu](mailto:nslavov@northeastern.edu).

This project was supported by funding by an [Allen Distinguished Investigator Award](https://alleninstitute.org/person/nikolai-slavov/) from the Paul G. Allen Frontiers Group to N.S., an NIGMS award R01GM144967 to N.S., NCI awards UG3CA268117 and UH3CA268117 to N.S., and a MIRA award R35GM148218 from the NIGMS of to N.S.


[Decode_article]: https://doi.org/10.1038/s41587-022-01389-w "Alternate RNA decoding results in stable and abundant proteins in mammals"
[decode_Code]: https://github.com/SlavovLab/decode "Decode data analysis pipeline, GitHub repository from the Slavov Laboratory"
